adr/graph-feature-flags.md

```diff
@@ -0,0 +1,291 @@
+If it does, the Router will execute the request with the feature flag configuration.
+If it does not, the Router will return a 400 Bad Request error.
+
+### Analytics and Metrics need to take into consideration Feature Flags
+
+Both Metrics and Tracing need to be aware of feature flags.
+Each entry needs to contain the feature flag that was used for the request so that one can filter by feature flag.
+It's a requirement to be able to understand metrics like error rates, latency, and throughput per feature flag.
+
+### The Studio should show all Feature Flags
+
+The Studio should show all feature flags that are available for a federated graph.
+
+### The Playground needs to be aware of Feature Flags
+
+On the Playground, it should be possible to see all available feature flags for a federated graph.
+The Playground should allow selecting a feature flag from a dropdown.
+The Playground should pre-select the default feature flag, which means that no feature flag is active.
+Based on the selected feature flag, the Playground should set the appropriate header or cookie.
+
+### Feature Flags must work for all types of requests
+
+- Subscriptions over WebSocket (header, cookie, extensions)
+- Queries and Mutations over HTTP (header, cookie, extensions)
+- Subscriptions over SSE (header, cookie, extensions)
+
+### Introspection needs to be aware of Feature Flags
+
+Depending on the selected feature flag, the introspection query should return the schema that is associated with the feature flag.
+
+### No security through obscurity
+
+Feature Flags are not a security feature.
+They are not hidden.
+Publishing a feature flag is a public action.
+
+If you need to publish a feature flag privately, create a dedicated federated graph with restricted access.
+
+### Feature Flags info is forwarded to the Subgraph
+
+The Router will forward the feature flag Header to Subgraphs automatically.
+
+### Feature Flags can be listed in the Studio and the CLI
+
+The Studio and the CLI should show all available feature flags for a federated graph.
+
+```shell
+wgc ff list # lists all feature flags and group feature flags
+```
+
+### Feature Flags can be disabled in the router.yaml
+
+Feature Flags can be associated with a federated graph using label matching.
+However, you might want to run different Routers in different networks with feature flags enabled or disabled.
+For this reason, it should be possible to disable feature flags in the router.yaml and allow to configure an allow list of regexes for feature flags.
+
+If feature flags are disabled in the router.yaml or a feature flag is not in the allow list, the Router will return a 400 Bad Request error.
+If feature flags are enabled and the allow list is empty, all feature flags are allowed.
+By default, feature flags are enabled and all feature flags are allowed.
+
+```yaml
+featureFlags:
+  enabled: true
+```
+
+### Feature Flags can be enabled or disabled using the CLI
```

Comment:
If a FF is disabled, and a client makes a request trying to use it, will the default sub-graph just get used?

And if the FF version of the sub-graph has some additional field, that doesn't exist in the default sub-graph, what happens then?

Reply:
In this case, the router would return a 400 bad request. Do you think this would be the expected behavior?

Reply:
I think that's fine. Just means that we need to be careful that our configuration on clients should be updated first (probably with our own Feature Flags), before the feature flag is disabled/removed on the WG side.